tgui/packages/tgui/components/NumberInput.tsx

```diff
@@ -0,0 +1,351 @@
+import { KEY } from 'common/keys';
+import { clamp } from 'common/math';
+import { BooleanLike, classes } from 'common/react';
+import {
+  Component,
+  createRef,
+  FocusEventHandler,
+  KeyboardEventHandler,
+  MouseEventHandler,
+  RefObject,
+} from 'react';
+
+import { AnimatedNumber } from './AnimatedNumber';
+import { Box } from './Box';
+
+const DEFAULT_UPDATE_RATE = 400;
+
+type NumberInputProps = {
```

Comment:
these should be just "props" and "state" unless we're working with multiple components in one file. this is to avoid type names getting overly complex and certainly resembling java (this kills the crab)